### ISSUE TITLE
论文“Ranked List Loss for Deep Metric Learning”复现的一点分享。

### ISSUE THREAD

**Issue by Qidian213** [michuanhaohao/reid-strong-baseline#13]

**Comment by Qidian213** [comment]
另外因为实验较少，未能确定实验效果是主要有rank-loss带来的，还是所设计的features ft, features fi分开做loss计算的 bnneck结构所带来的。希望能得到更多指点。

**Comment by michuanhaohao** [comment]
@Qidian213 非常感谢实验，你这个用了所有的trick吗，用了所有的trick在不rerank的条件下我们的baseline是86.4（76.4）的性能，貌似并没有涨点的样子。我看你的log里面只用了label smoothing

**Comment by Qidian213** [comment]
Warm up learning rate 
Random erasing augmentation = 0.5
Label smoothing =  on
Last stride = 1
看配置文件，应该是用了这四种，上面实验主要是把损失函数替换了,主要也是为了在相同方式下与triplet loss的作对比。另外也是因为实验太少，未能顾忌到。

**Comment by michuanhaohao** [comment]
@Qidian213 所以BNNeck也没用是吧，如果没用应该是涨点了的

**Comment by Qidian213** [comment]
是用了BNNeck的，我前面说了 “保留BN layer 与fc layer 并使用 features fi做测试” 。昨晚又跑了次训练测试，隐约的觉得 rank-loss权重应该小点，于是 最后Loss 变成了 loss = xent_loss + 0.1*ranked_loss. 然后认为负样本距离可以不达到 alpha, 正样本需要达到 alpha-margin内，于是尝试新参数 alpha =2.0, margin = 1.6, T = 1.0.

```
      IF_LABELSMOOTH: 'on'
      IMS_PER_BATCH: 64
      FEAT_NORM: 'no'
      NECK = 'bnneck'
      TEST.RE_RANKING 'yes'
      mAP: 88.3%
      Rank-1  :90.1%
      Rank-5  :94.4%
      Rank-10 :96.2%
```

```
      IF_LABELSMOOTH: 'on'
      IMS_PER_BATCH: 64
      FEAT_NORM: 'no'
      NECK = 'bnneck'
      mAP: 72.6%
      Rank-1  :87.0%
      Rank-5  :94.0%
      Rank-10 :95.9%
```

```
      IF_LABELSMOOTH: 'on'
      IMS_PER_BATCH: 64
      FEAT_NORM: 'yes'
      NECK = 'bnneck'
      TEST.RE_RANKING 'yes'
      mAP: 88.3%
      Rank-1  :89.6%
      Rank-5  :94.5%
      Rank-10 :96.1%
```

```
      IF_LABELSMOOTH: 'on'
      IMS_PER_BATCH: 64
      FEAT_NORM: 'yes'
      NECK = 'bnneck'
      mAP: 75.6%
      Rank-1  :85.9%
      Rank-5  :94.0%
      Rank-10 :96.2%
```
完全使用你在配置文件里设置的learning-rate ,max_epochs。 最后发现跑完 120个epochs后仍在收敛，在测试集rank-1还在提升，继续训练应该能达到更好的效果。
测试发现 “FEAT_NORM: 'no'”比“FEAT_NORM: 'no'” rank-1 高了1.1%。map却低了 3% 。 我认为当做类内距离小于alpha-margin后不再继续优化，会导致类内特征分布占据比能继续优化的更大的类内超球空间，在归一化后所有类的小超球分布在一个大的超球上，导致大的超球分布较为拥挤一些，取消归一化后不受超球面的限制，可以分布到更大的空间中，类间分布更清晰。还未能想清楚map低的原因。
不知道这种想法对不对，希望能得到指点。

**Comment by Qidian213** [comment]
在超参数中，loss = xent_loss + 0.1*ranked_loss. 的0.1和 T 值都会影响到负样本的学习率。这样超参数较多，调参会繁琐点。

**Comment by michuanhaohao** [comment]
@Qidian213 目前来看结论是rank loss和triplet loss可以得到差不多的性能（略微低一点），如果参数调到比较好的话也许可以超过。另外纠正一个feat norm参数的理解错误，这个参数不影响训练过程，只是测试的时候使用BN前的f_t还是BN后的f_i而已，训练的时候约束的都是BN前的feature。可以训一个有无rank loss的对比实验，看看rank loss是否能涨点

**Comment by Qidian213** [comment]
这两天又有些空，补充点数据。eg: Train: F_t 指rank loss 所使用的特征为F_t。
实验一, 只使用Rank loss或者 ID loss做训练。
        只用 rank_loss
```
        Train: F_t 
        Test : F_t
        map : 55.7%
        rank1: 75.0%
```

```
        Train: F_i
        Test : F_i
        map : 58.9%
        rank1: 76.4%
```

只使用ID loss

```
        map : 72.5%
        rank1: 84.4%
```
实验二，使用ID loss + W* Rank loss 的方式

```
        Train_epochs: 140
        Train: F_t

        Test : F_i
        map : 75.9%
        rank1: 86.3%

        Test : F_t
        map : 72.9%
        rank1: 87.3%
```

```
        Train_epochs: 80
        Train: F_i

        Test : F_i
        map : 76.4%
        rank1: 87.6%
```

**Comment by michuanhaohao** [comment]
结论可以得到Rank loss是有用的，但是在BN后面的feature训和测更好，对于BN这个东西，越来越迷了

**Comment by michuanhaohao** [comment]
@Qidian213 有空的话能不能给一下最后这组实验，market的结果，想要看看是不是能得到同样的结论

**Comment by Qidian213** [comment]
没问题的，但是需要晚几天。

**Comment by Qidian213** [comment]
实验二，使用ID loss + W* Rank loss 的方式  Market1501数据集
```
        Train_epochs: 110
        Train: F_t

        Test : F_i
        map : 85.3%
        rank1: 93.9%

        Test : F_t
        map : 70.1%
        rank1: 86.4%
```

```
        Train_epochs: 120
        Train: F_i

        Test : F_i
        map : 86.4%
        rank1: 94.7%
```
虽然整体看 BN后面的feature训和测更好，但相比论文中并没有明显提升，不知可不可以判定为rank loss有效？ 疑惑的是训练 BN层前的特征， 使用 F_t 测试与 F_i 性能相差太大，与在DukeMTMC-reid上表现不符。

**Comment by XinshaoAmosWang** [comment]
Our code is released recently: 
https://github.com/XinshaoAmosWang/Ranked-List-Loss-for-DML

I am gonna conduct experiments on ReID, e.g., MARS, with RLL shortly.

**Comment by Qidian213** [comment]
@XinshaoAmosWang  Thanks for sharing, I also released my code based on your paper, some may different and results  better.  https://github.com/Qidian213/Ranked_Person_ReID , look forward to discussion with you.

**Comment by michuanhaohao** [comment]
Thank you both!

**Comment by anonymoussss** [comment]
这个Ranked List Loss的pytorch库复现不出来 有没有能复现出来的dalao啊

**Comment by XinshaoAmosWang** [comment]
#### Our code can be downloaded from
https://www.dropbox.com/sh/bg8pdxlxik1idqa/AADjK3XAAW4ymOPxxQAaQXyka?dl=0

#### What language I used: 
* I used C++ in the caffe framework
#### Where is the loss function defined:
* CaffeMex_RLL_GR_V03_Simp/src/caffe/layers/distance_weighted_valid_pair_loss_V12_layer.cpp

If you have any difficulties in reproducing our t-pami results, please feel free to let me know (I have added some new experiments for t-pami without updating this dropbox repository yet. But the changes are not many, therefore, please feel free to use the code). 
https://ieeexplore.ieee.org/document/9385896

Please note the Terms and conditions when using this code. 
#### Terms and conditions: 

* The download and use of the code is released for academic research only and it is free to researchers from educational or research institutes for non-commercial purposes.
When downloading the code you agree to (unless with expressed permission of the authors): not redistribute, or commercial usage of this code in any way or form, either partially or entirely.

* If using this code, please cite the following paper: https://ieeexplore.ieee.org/document/9385896
@article{wang2021rankedlistloss_tpami,
  title={Ranked List Loss for Deep Metric Learning},
  author={Wang, Xinshao and Hua, Yang and Kodirov, Elyor and Robertson, Neil M}, journal={IEEE Transactions on Pattern Analysis and Machine Intelligence},
  year={2021},
  publisher={IEEE}
}
@inproceedings{Wang_2019_CVPR,
  author = {Wang, Xinshao and Hua, Yang and Kodirov, Elyor and Hu, Guosheng and Garnier, Romain and Robertson, Neil M.},
  title = {Ranked List Loss for Deep Metric Learning},
  booktitle = {The IEEE Conference on Computer Vision and Pattern Recognition (CVPR)},
  month = {June},
  year = {2019}
}


Best wishes.
